### DISCUSSION TITLE
Update `MartyrologyDef`

### DISCUSSION THREAD

**Discussion by tukusejssirs** [romcal/romcal#676] [Ideas]
### Description of the issue

As there are many types of celebrations (not from the liturgical POV), we should require different properties for each of them.

I suggest creating a type for each celebration type.

For romcal needs, I propose to define the following celebration types:

- personal celebrations (`PersonalCelebration`);
- event celebrations (`EventCelebration`);
- dedication celebrations (`DedicationCelebration`).

There might be more celebration types, we can add them when we find them. :wink:

Moreover, there needs to be a special type (albeit not a celebration type) that would be used for collective names, i.s. to define what persons are part of a collective (including `companions`). For the sake of simplicity, from now on when I say _celebration type_, I also mean collective names in this sense.

Finally, we might want to consider adding temporal celebrations (Proper of Times), so that we have all celebration metadata defined in a single place. For that, it might be a good idea to define the variable dates in a codeless manner (i.e. just using POJO properties).

### Organisation point of view

From the organisation point of view, we can keep all celebration types (and collective names) in a single file (`martyrology.ts`) or we might decide to move each celebration type (as well as collective names) to separate files. I have no strong opinion on this. For the sake of this issue (in its draft form), I presume we want to keep it as a single file.

### Proposal

I propose the following definitions for each celebration type (consider these types as drafts):

```typescript
/**
 * Celebration type
 *
 * @remarks A special division of the celebrations. We also include collectives as a special celebration type.
 */
enum CelebrationType {
  /** Collective name */
  COLLECTIVE = 'collective',

  /** Dedication celebration */
  DEDICATION = 'dedication',

  /** Event celebration */
  EVENT = 'event',

  /** Personal celebration */
  PERSON = 'person',
}

/** Sex of the collective */
enum CollectiveSex {
  /** Purely female collective */
  female = 'FEMALE',

  /** Purely male collective */
  male = 'MALE',

  /** Mixed-sex collective */
  mixed = 'MIXED',
}

/** Collective name */
type CollectiveName = {
  /**
   * Number of persons in the collective
   *
   * @remarks It could be set to `many` if the number is unknown.
   */
  count: SaintCount;

  /** Collective name */
  name: string;

  /**
   * Sex of the collective
   */
  sex: Sex;

  /** Celebration type */
  type: CelebrationType.COLLECTIVE;

  /** Persons who make up the collective */
  persons?: string[]
}

/** Dedication celebration */
type DedicationCelebration = {
  /** Church which dedication is defined in this object */
  church: name

  /** Dedication date */
  date?: SaintDateDef,

  /** Celebration type */
  type: CelebrationType.DEDICATION;
}

/** Event celebration */
type EventCelebration = {
  /** Celebration name */
  name: string;

  /** Celebration type */
  type: CelebrationType.EVENT;
}

/** Personal celebration */
type PersonalCelebration = {
  /** Beatification metadata */
  beatification?: {
     /** Beatification date */
    date?: SaintDateDef,
     /** Beatification place */
    place?: string
  },

  /** Birth metadata */
  birth?: {
    /**
     * Birth date
     *
     * @remarks It could be defined as a number (year), a string (in `YYYY-MM` or `YYYY-MM-DD` formats), or an object describing date range, multiple possible date, or a century.
     */
    date?: SaintDateDef,

    /** Whether the birth date is approximate */
    isApproximate?: boolean,

    /** Birth place */
    place?: string
  },

  /** Canonisation metadata */
  canonisation?: {
     /** Canonisation date */
    date?: SaintDateDef,
     /** Canonisation place */
    place?: string
  },

  /**
   * The canonization level of a person
   */
  canonizationLevel: CanonizationLevel;

  /** Death metadata */
  death?: {
    /**
     * Death date
     *
     * @remarks It could be defined as a number (year), a string (in `YYYY-MM` or `YYYY-MM-DD` formats), or an object describing date range, multiple possible date, or a century.
     */
    date?: SaintDateDef,

    /** Whether the death date is approximate */
    isApproximate?: boolean,

    /** Death place */
    place?: string
  },

  /** Dedication metadata */
  dedication?: {
     /** Dedication date */
    date?: SaintDateDef,
  }

  /**
   * First name
   */
  name: string;

  /**
   * Sex of the person
   */
  sex: Sex;

  /**
   * Last name
   */
  surname?: string;

  /**
   * Titles by which the person is venarated
   *
   * @remarks Not all titles are used at all places/calendars, however, all titles by which the person is verated at least at one place should be listed here.
   */
  titles?: {
    /** Common titles */
    common?: Title[],

    /** Patronage titles */
    patronage?: PatronageTitle[]
  };

  /** Celebration type */
  type: CelebrationType.PERSON;
}

/**
 * Martyrology definition
 *
 * @remarks This type is used for any item in the celebration catalog.
 */
type MartyrologyDef = CollectiveName | DedicationCelebration | EventCelebration | PersonalCelebration
```

### Notes

1. I think that `hideCanonizationLevel` is not necessary after we define which properties are required per celebration type, as it was used by non-personal celebrations (basically event celebrations) only. Therefore, I propose to remove `hideCanonizationLevel`.
2. I think that `hideTitles` is not necessary too, as we should always choose (unless inherited via calendar inheritance) which titles should be included in a celebration within a calendar, as some persons are venerated by different common/patronage titles in different particular calendars.
3. We might also want to move the martyrology item ID to the property itself as an `id`.
4. [`PersonalCelebration`] While `name` is (according to the TSDoc) deprecated ATM, I think we should not remove it, but we should change what value it should have. IMHO we should split it to `name` and `surname` and define new IDs for these. Moreover, `name` should have an enum type (or deeply readonly variable → `const firstNames = {...} as const`), but we need to keep in mind that some names in one locale can have multiple forms in the other (and vice versa). Then, we should localise the names and surnames separately (see #154). I expect some issues with this (some names/surnames are not always localised the same way, e.g. Spanish vs Portuguese surnames), but I still think that at least 90 % of time this will help us more than cause troubles.
5. [`PersonalCelebration`] I am also considering making `surname` required. The only persons that do not have surnames are the twelve Apostles, Jesus (we could use _Christ_), Our Lady (which is defined more than three dozen times). We could define the surnames for these people.
6. [`PersonalCelebration`] Jesus, Our Lady, St Joseph and St John the Baptist have multiple celebrations catalogued, but no personal definition. IMHO we should create a single personal celebration definition in `martyrology` with a property that would list (or link by ID) all the celebrations.
7. [`PersonalCelebration`] I also think we should all the (liturgical) titles of all persons, both common (priest, martyr, etc) and patronage titles (see #221, #201, and also some other issue/discussion where Etienne and I discussed the patronage title definition type). IMHO all patronage places (patron of what is a particular person) should be (similarly to `name` and `surname`) should be listed in an enum (or deeply readonly variable) and referenced in the items of `martyrology.ts`.
8. [`PersonalCelebration`] I have split the common titles from the patronage titles into subproperties, as IMHO patronage titles should be defined as a property, not as a string like `COPATRON_OF_EUROPE`. Note that the `PatronageTitle` type still needs to be defined.
9. [`PersonalCelebration`] We might want to remove `PersonalCelebration.canonizationLevel`, as we can compute it fairly easily by changed whether `canonisation` property exists (if yes, it is a `saint`; if not, it is a `blessed`). However, I am not opposed to keeping it.
10. [`PersonalCelebration`] We might want to add additional properties I might have forgotten.
11. I think we should also list all places in some enums (or deeply readonly variables), e.g. `Continents`, `Subcontinents`, `Regions`, `Countries`, `Dioceses`, `Cities` (including villages if necessary, but only those mentioned in national Calendar Propers or at least in ordos). Then, we should reference them in `martyrology.ts` and localise them in the l10n files.
12. I am also thinking about updating the definition of `SaintDateDef`.
13. [`CollectiveName`] How can we type `CollectiveName.persons` so that only personal celebration IDs from `martyrology.ts` can be used? :thinking:
14. [`DedicationCelebration`] We should list all churches (as buildings) that we mention/use in romcal. Then, localise them. We should also define the metadata of churches, like `type` (cathedrals, basilicas, domes, churches, chapels), `city` (and via `city` also `country`, `subcontinet`, `continent`, and optional `region`), and maybe some other metadata.
15. [`PersonalCelebration`] We also should add metadata objects for the following information (I am not sure whether we should create a single metadata object for deacon/priest/bishop/cardinal/pope metadata or several ones; however, it would be awesome to only allow defining some properties if a respective common title is defined, e.g. priest metadata can be defined only if `Title.Priest` is defined in `PersonalCelebration.titles.common`):

  - deacon metadata:
    - consecration date and place;
  - priest metadata:
    - consecration date and place;
  - bishop metadata
    - consecration date and place;
    - diocese;
    - motto;
  - cardinal metadata
    - level (`deacon`, `priest`, `bishop`);
    - creation date and place;
    - resignation date;
    - titular church (cardinal deacons: `deaconry`; cardinal priests: `title` or `title _pro hac vice_ (for this occasion); cardinal bishop: `suburbicarian diocese`);
  - pope metadata
    - election date and place;
    - inauguration date and place;
    - [required] papal name;
    - resignation date;
    - motto;
  - religious metadata
    - consecration date and place;
    - religious name;
    - order/congregation/institute (should we name the property as `community`?):
      - I think we should create an object with metadata of the religious orders/congregations/institute, like official and unofficial name, founder, establishment year, cancellation year, type (order/congregation/institute);
  - martyr metadata:
    - who killed them (prosecutor?);
    - martyrdom type:
      a. `in odium fidei`: being killed due to hatred of faith;
      b. `ex aerumnis carceris`: dying in prison for the faith;
      c. `in defensum castitatis`: being killed while defending virginal purity;
      d. `per testimonium caritatis fortis`: being killed as a result of heroic charity;
      e. `ex acertatibus et vexationibusque pro fidei quibus pertulit`: dying as a result of physical or moral violence endured for the sake of the faith.

16. I think we should also add an optional `notes` property to each celebration type. While it would not be used by romcal, it might be actually useful to add some additional metadata.

17. I think it would be nice to add another public `getter` function to the library that could be used by client apps.

**Comment by TobiTenno** [comment]
This is overly expansive, and should be better subdivided to have individual parts we can discuss. having 17 opinionated pieces that you want to add makes this hard to discuss moving forward on it. Several of these can probably be their own issue, but any new fields should be separated from any fields you want to remove unless there are pairs that are explicitly entertained. I agree in concept, but we need to make these thin vertical slices that we can move forward on instead of more and more pieces that will collectively hold us from putting smaller pieces into place.

i.e. no. 17 doesn't need to be blocked by adding all of the no. 15 annotations.

**Comment by tukusejssirs** [comment]
# TODO

- [ ] rename `WEEK_{1,2}` and `YEAR_{1,2}` to `WEEK_{I,II}` and `YEAR_{I,II}`, respectively;
- [ ] automatically update romcal versions in `README.md`:
  - `dev` versions: `romcal@3.0.0-dev.7`/`romcal@dev`;
  - `dev` versions: `romcal@1.3.1-beta.0`/`romcal@beta`;
  - `latest` versions: `romcal@3.0.0`/`romcal`/`romcal@latest`;
- [ ] in `README.md`: if we are to mention both `npm` and `yarn`, how about mentioning `pnpm`?
- [ ] in `README.md`: fix some grammar (not yet finished):
- [ ] remove the `dedication_of_consecrated_churches_on_*` l10n keys (see https://github.com/romcal/romcal/issues/136#issuecomment-1701541394);
- [ ] fix `gabriel_taurin_dufresse_bishop` key name (see https://github.com/romcal/romcal/pull/458/files#r1331061512);
- [ ] commit `pnpm-lock.yaml`:
    - decide whether we should remove `package-lock.json` or update it in GitHub CI from `pnpm-lock.yaml` (a PNPM command);

  ```diff
  diff --git a/README.md b/README.md
  index 722ab09..5b72849 100644
  --- a/README.md
  +++ b/README.md
  @@ -84,7 +84,7 @@ The correct entry points are already configured in the package.json so there sho

   ### Calendars installation

  - The romcal library only include the General Roman Calendar (GRC), and the Proper of Time. By default, there is no other calendars, neither translation (even in English) nor extra martyrology metadata.
  +The romcal library only includes the General Roman Calendar (GRC), and the Proper of Time. By default, no other calendar or translation (even in English) nor extra martyrology metadata.

   The complete GRC, and any other particular calendar (for a country, a region or a diocese) are available as separated plugins, that contain a bundle of the calendar data, localizations, and a martyrology catalog (containing extra metadata).
  ```

- add `.idea`;
- sort `package.json`, esp `scripts`;
- consider using `.yml` extension instead of `.yaml`;

## romcal plugins

I want to discuss the way we implement plugins in romcal.

IMO user should be able to add multiple plugins at once and all of them should be considered during the calendar generation.

In the future, I think we should have plugins for the following tasks (some tasks could be done in a single plugin):

- define additional items or update existing ones in the celebration catalogue (`martyrology.ts`):
   - when we add celebrations in plugins, we need to be able to add items to the celebration catalogue;
   - this might be useful for debugging purposes (before merging the changes to romcal);
- define additional calendar (-s):
   - use case: parish calendars with the parish patron, parish church patron, dedications of churches and chapels, patrons of the filial churches (note: I am not sure how are _filial_ churches (as buildings) or _filias_ (churches are a community) called in English; maybe @TobiTenno knows);
   - this might be useful for debugging purposes (before merging the changes to romcal);
- define additional celebrations or modify existing celebrations in a calendar:
   - this might be useful for debugging purposes (before merging the changes to romcal);
- add a way to select which celebration will be celebrated on a particular day (feria vs optional memorials):
   - this would be useful for planning the liturgical celebrations in a particular parish, or even better: define a celebration/formular for a specific Holy Mass (as different Masses on the same date in the same parish might be celebrated Holy Masses with different formular);
   - don’t forget about votive Masses (incl those to saints/blesseds inscribed in the _Martyrologium Romanum_ on that day):
      - OT: this is why I think we should have more persons/celebrations defined in the celebration catalogue (`martyrology.ts`) than only those that are to be used directly by romcal;
- generate a calendar that would include a national calendar and either all or only selected dioceses of that country:
   - this is how ordos do it (at least in Hungary and Slovakia): they list all celebrations celebrated on the national level and if there are different celebrations celebrated on the diocesan level (e.g. dedications or patronages), they are listed as the last celebration of a day;
- define additional locales:
   - this might be useful for debugging purposes (before merging the changes to romcal);
- add additional strings or update existing ones in `.names` locales:
   - when we add celebrations in plugins, we need to be able to localise them;
   - this might be useful for debugging purposes (before merging the changes to romcal);

## update mart (comment)

> This is overly expansive, and should be better subdivided to have individual parts we can discuss.

@TobiTenno, okay, I’ll convert this issue to discussion.

IMHO, we should discuss and try to get some consensus on how this object (an item in the celebration catalogue) should look like in the end (at least from higher POV), then we should try to create some plan/points on how to achieve the expansion of celebration metadata in a more iterative way.

> having 17 opinionated pieces that you want to add makes this hard to discuss moving forward on it.

I agree these suggestions are very opinionated, but they come from my experience with romcal and liturgical calendars in general. From localisation point of view, my biggest issue is #154: I wish romcal could generate celebration names (titles) from individual pieces, as ATM there is much duplication and human errors occurred quite often to either add some typos or use a different capitalisation. All of this can be automated, _as long as we have the metadata in place first_.

And yes, most of the metadata is used only by some locales, other metadata is not used at all (but they might be a nice addition).

> Several of these can probably be their own issue, but any new fields should be separated from any fields you want to remove unless there are pairs that are explicitly entertained.

Yeah, we should create issues for each separate problem, so that we could more in a more iterative manner, possibly the discuss them separately

> I agree in concept, but we need to make these thin vertical slices that we can move forward on instead of more and more pieces that will collectively hold us from putting smaller pieces into place.

:100: agreed.

> i.e. no. 17 doesn't need to be blocked by adding all of the no. 15 annotations.

Yeah, the list was never meant to be ordered in the implementation order. :wink: I agree that the number 17 is a small feature to add (it can be done right away). :wink:
